jfix-stdlib-concurrency/src/main/kotlin/ru/fix/stdlib/concurrency/events/EventReducer.kt

```diff
@@ -0,0 +1,76 @@
+package ru.fix.stdlib.concurrency.events
+
+import mu.KotlinLogging
+import java.util.concurrent.*
+
+
+/**
+ * EventReducer invokes given [handler] when [handle] function was invoked.
+ * If [handle] function was invoked (one or many times) during [handler] working,
+ * [handler] will be invoked after completing and so on.
+ *
+ * Therefore if [handle] function was invoked 100 times with breaks in 1 millisecond,
+ * and [handler]'s work takes 70 milliseconds,
+ * then [handler] wasn't invoked 100 times, but 3 times, consistently.
+ * */
+class EventReducer(
+        private val handler: () -> Unit,
+        private val shutdownCheckPeriodMs: Long = 1_000,
+        private val awaitTerminationPeriodMs: Long = 60_000
+) : AutoCloseable {
+    private val eventReceivingExecutor = Executors.newSingleThreadExecutor()
+
+    private val awaitingEventQueue = ArrayBlockingQueue<Any>(1)
+
+    fun handle() = synchronized(awaitingEventQueue) {
+        if (awaitingEventQueue.isEmpty()) {
+            awaitingEventQueue.put(Any())
+        }
+    }
+
```

Comment:
Right know EventReducer `reduce` empty events. User can not provide custom reducing logic at all.  If you are building generic solution that can be used by others I suggest you to add ability for api to accept and reduce actual user events objects:
```
// frequently invoked for each new event
fun handleEvent(event: Any?)

// frequently invoked for each new event
// receive accumulator and event, returns new accumulator
// for first event reducer will be invoked with null accumulator value: reducer(null, event)
// E.g. we receive 3 events: evetn1, event2, event3, then reducer will be invoked in this way: 
// var acc = reducer(null, event1)
// acc = recuer(acc, event2)
// acc = reducer(acc, event3)
val reducer: (Any?, Any?) -> Any? 

// rarely invoked at the end of butch of events
// receives an accumulator value returned by reduce function
val handler: (Any?) -> Unit
```

for our case in Zk DJM we can provide default `last win reducer`
```
reducer = acc, event -> event
```
This way we confirm open-close principal and allow our component to be extendable

Reply:
fixed, EventReducer now is generic. I think its better to use generic types instead of `Any`, it allows don't cast types in users code